lib/Target/SimFHE/SimFHEEmitter.cpp

```diff
@@ -0,0 +1,179 @@
+  os.unindent();  // exit internal for loop
+  os << kMainPostlude << "\n";
+  return success();
+}
+
+LogicalResult SimFHEEmitter::printOperation(func::FuncOp funcOp) {
+  os << "def " << funcOp.getName() << "(";
+  os << heir::commaSeparatedValues(funcOp.getArguments(),
+                                   [&](Value value) { return getName(value); });
+  if (funcOp.getNumArguments() > 0) os << ", ";
+  os << "scheme_params : params.SchemeParams):\n";
+  os.indent();
+  os << "stats = PerfCounter()\n";
+  for (Block &block : funcOp.getBlocks()) {
+    for (Operation &op : block.getOperations()) {
+      if (failed(translate(op))) {
+        return failure();
+      }
+    }
+  }
+  os << "return stats\n";
+  os.unindent();
+  os << "\n";
+  return success();
+}
+
+LogicalResult SimFHEEmitter::printOperation(func::ReturnOp op) {
+  // return is handled via accumulating stats; nothing to do
+  return success();
+}
+
+#define UNARY_EMIT(OPTYPE, GETTER, FUNCNAME)                 \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {   \
+    std::string input = getName(op.GETTER());                \
+    os << "stats += evaluator." FUNCNAME "(" << input        \
+       << ", scheme_params.arch_param)\n";                   \
+    os << getName(op.getResult()) << " = " << input << "\n"; \
+    return success();                                        \
+  }
+
+#define UNARY_EMIT_DROP(OPTYPE, GETTER, FUNCNAME)                   \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {          \
+    std::string input = getName(op.GETTER());                       \
+    os << "stats += evaluator." FUNCNAME "(" << input               \
+       << ", scheme_params.arch_param)\n";                          \
+    os << getName(op.getResult()) << " = " << input << ".drop()\n"; \
+    return success();                                               \
+  }
+
+#define BINARY_EMIT(OPTYPE, GETTER, FUNCNAME)                \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {   \
+    std::string input = getName(op.GETTER());                \
+    os << "stats += evaluator." FUNCNAME "(" << input        \
+       << ", scheme_params.arch_param)\n";                   \
+    os << getName(op.getResult()) << " = " << input << "\n"; \
+    return success();                                        \
+  }
+
+// Since we only care about the cost, add/subtract are treated the same
+BINARY_EMIT(ckks::AddOp, getLhs, "add");
+BINARY_EMIT(ckks::AddPlainOp, getLhs, "add_plain");
+BINARY_EMIT(ckks::SubOp, getLhs, "add");
+BINARY_EMIT(ckks::SubPlainOp, getLhs, "add_plain");
+BINARY_EMIT(ckks::MulOp, getLhs, "multiply");
+BINARY_EMIT(ckks::MulPlainOp, getLhs, "multiply_plain");
+UNARY_EMIT(ckks::NegateOp, getInput, "multiply_plain");
+UNARY_EMIT(ckks::RotateOp, getInput, "rotate");
+LogicalResult SimFHEEmitter::printOperation(ckks::RelinearizeOp op) {
+  std::string name = getName(op.getInput());
+  os << "stats += evaluator.key_switch(" << name << ", " << name
```

Comment:
Why does key switch duplicate the argument, and why can't this be an instance of the binary op case?